### DISCUSSION TITLE
How to create custom chain of credential providers

### DISCUSSION THREAD

**Discussion by avishay28** [aws/aws-sdk-go-v2#2894] [Q&A]
Hey,
How can I mimic this functionally from v1 sdk?

```go

import (
	"github.com/aws/aws-sdk-go/aws/credentials"
	"github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds"
)

...

func GetCreds(accessKey, secretKey, token, region string) (*credentials.Credentials, error) {
	creds := credentials.NewChainCredentials([]credentials.Provider{
		&credentials.StaticProvider{Value: credentials.Value{
			AccessKeyID:     accessKey,
			SecretAccessKey: secretKey,
			SessionToken:    token,
		}},
		&credentials.EnvProvider{},
		&credentials.SharedCredentialsProvider{Filename: "", Profile: ""},
		&CustomProvider1{region: region, ExpiryWindow: 10 * time.Second},
		&ec2rolecreds.EC2RoleProvider{
			ExpiryWindow: 10 * time.Second,
			Client: ec2metadata.New(session.New(&aws.Config{
				Region: aws.String(region),
			})),
		},
		&CustomProvider2{Foo: "bar"},
	})

	_, err := creds.Get()
	if err != nil {
		return nil, err
	}

	return creds, nil
}
```

I couldn't find any way to do that in v2, it looks like a lof ot the credentials logic is hidden, I don't have any controll on the chain order.
Now I need to use `LoadDefaultConfig` which forcing the order of the providers.
I could try to create a customer provider which chains providers together and try to `Retreive` them in order, however it means I need to write in hand providers like `SharedCredentialsProvider` or `EnvProvider` which are not exist in v2.

As a general feedback- I feel like there's a bit of mess in everything that related to the way we working with credentials in v2 sdk. Some logic belongs to `credentials` package, some to `aws` package (like the `Credentials` struct or `CredentialsCache`, like.. why `CredentialsCache` logic is not part of `credentials` package?), some to `config`.
It was much easier to work with when everything was under `github.com/aws/aws-sdk-go/aws/credentials`

**Comment by codelinx** [comment]
can you outline more succinctly:
- what you are trying to accomplish?
- what was the previous process + outcome vs the new outcome? 
- what would you like to get from the feedback on here?


i am no expert in go or the SDK, but it seems packages were organized into new logic units for user experience and better descriptors of the functionality. 🤷
